lib/summarize.py

```diff
@@ -1,65 +1,75 @@
 # SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
 # SPDX-License-Identifier: AGPL-3.0-or-later
-"""A recursive summarize chain
-"""
 
 from typing import Any
-
+from langchain.schema.messages import HumanMessage, SystemMessage
 from langchain.prompts import PromptTemplate
 from langchain.schema.prompt_template import BasePromptTemplate
-from langchain.text_splitter import CharacterTextSplitter
 from langchain_core.runnables import Runnable
+from langchain.text_splitter import RecursiveCharacterTextSplitter
 
 
 class SummarizeProcessor:
-    """
-    A summarization chain
-    """
+    runnable: Runnable
+    text_splitter: RecursiveCharacterTextSplitter
 
     system_prompt: str = "You're an AI assistant tasked with summarizing the text given to you by the user. Use a bullet list to summarize. Make sure to cover all topics of the text in your summary. Output only the summary without quotes, nothing else, especially no introductory or explanatory text."
     user_prompt: BasePromptTemplate = PromptTemplate(
-        input_variables=["text"],
+        input_variables=["input"],
         template="""
-Summarize the following text. Detect the language of the text. Use the same language as the one you detected. Here is the text:
+Summarize the following text. Detect the language of the text. Use the same language as the one you detected. Also, do not mention the language you used explicitly. Here is the text:
 
----
-{text}
----
+Text to summarize:
 
- Also, do not mention the language you used explicitly.
-Here is your summary in the same language as the original text:
-        """
+{input}"""
     )
 
+    merge_prompt: BasePromptTemplate = PromptTemplate(
+        input_variables=["input"],
+        template="""
+Combine these summaries into one coherent summary. Preserve the most important information while making it concise and easy to understand.
+The final summary should be in the same language as the original summaries. Detect the language of the original summaries. Use the same language for the final summary as the one you detected. Only output the final summary, no explanations or additional text. Do not mention the language used.
 
-    runnable: Runnable
-    n_ctx: int = 8000
+Summaries to combine:
+{input}
+"""
+    )
 
-    def __init__(self, runnable: Runnable, n_ctx: int = 8000):
+    def __init__(self, runnable: Runnable, chunk_size):
         self.runnable = runnable
-        self.n_ctx = n_ctx
+        self.text_splitter = RecursiveCharacterTextSplitter(
```

Comment:
the default separators `["\n\n", "\n", " ", ""]` might not be ideal. What do you think about including the previous separators here? `'\n\n|\\.|\\?|\\!'` (not sure if there should be two `\\`)